Pluggable monitors (#6)
arduino/tooling-rfcs: RFCs/0004-pluggable-monitor.md

```JSON
{
  "event": "port_closed",
  "message": "serial port disappeared!"
}
```

or

```JSON
{
  "event": "port_closed",
  "message": "lost TCP/IP connection with the client!"
}
```

#### CLOSE command

The `CLOSE` command will close the currently opened port and close the TCP/IP connection used to communicate with the Client/IDE. The answer to the command is:

```JSON
{
  "event": "close",
  "message": "ok"
}
```

or in case of error

```JSON
{
  "event": "close",
  "error": true,
  "message": "port already closed"
}
```

#### QUIT command

The `QUIT` command terminates the monitor. The response to `QUIT` is:

```JSON
{
  "eventType": "quit",
  "message": "OK"
}
```

after this output the monitor exits. This command is supposed to always succeed.

#### Invalid commands

If the client sends an invalid or malformed command, the monitor should answer with:

```JSON
{
  "eventType": "command_error",
  "error": true,
  "message": "Unknown command XXXX"
}
```

#### Reference implementations

TODO...

> ### Integration with `arduino-cli` and core platforms

Reviewer: ```suggestion
### Integration with Arduino CLI and core platforms
```

In this section we will see how monitors are distributed and integrated with Arduino platforms.

#### Monitor tool distribution

The monitor tools must be built natively for each OS and the CLI should run the correct tool for the running OS.

> The distribution infrastracture is already available for platform tools, like compilers and uploaders, through `package_index.json` so, the most natural way forward is to distribute also the monitor tools in the same way.
> 3rd party developers should provide their monitor tools by adding them as resources in the `tools` section of `package_index.json` (at the `packages` level).
> 
> Let's see how this looks into a `package_index.json` example:

Reviewer: ```suggestion
The distribution infrastructure is already available for platform tools, like compilers and uploaders, through [the Arduino package index](https://arduino.github.io/arduino-cli/latest/package_index_json-specification) so, the most natural way forward is to distribute also the monitor tools in the same way.
3rd party developers should provide their monitor tools by adding them as resources in the `tools` section of their package index (at the `packages` level).

Let's see an example of adding a monitor tool to a package index:
```
I think it's more clear to refer to it as "Arduino package index" rather than "`package_index.json`" because only the primary official package index has this filename but this system applies to all package indexes. This is the approach taken in the Arduino CLI documentation, including the package index specification.

```diff
{
  "packages": [
    {
      "name": "arduino",
      "maintainer": "Arduino",
      "websiteURL": "http://www.arduino.cc/",

      "platforms": [
        ...
      ],
      "tools": [
        {
          "name": "arm-none-eabi-gcc",
          "version": "4.8.3-2014q1",
          "systems": [ ... ]
        },
        {
          "name": "ble-discovery",
          "version": "1.0.0",
          "systems": [ ... ]
        },
+       {
+         "name": "ble-monitor",  <--- Monitor is distributed as a TOOL
+         "version": "1.0.0",
+         "systems": [
+           {
+             "host": "x86_64-pc-linux-gnu",
+             "url": "http://example.com/ble-mon-1.0.0-linux64.tar.gz",
+             "archiveFileName": "ble-mon-1.0.0-linux64.tar.gz",

> +             "checksum": +SHA-256:0123456789abcdef0123456789abcdef0123456789abcdef",

Reviewer: ```suggestion
+             "checksum": "SHA-256:0123456789abcdef0123456789abcdef0123456789abcdef",
```

+             "size": "12345678"
+           },
+           ...
+         ]
+       }
      ],
    }
  }
}
```

> In this case we are adding an hypotetical `ble-monitor` version `1.0.0` to the toolset of the vendor `arduino`. From now on, we can uniquely refer to this monitor with the pair `PACKAGER` and `MONITOR_NAME`, in this case `arduino` and `ble-monitor` respectively.

Reviewer: ```suggestion
In this case we are adding an hypothetical `ble-monitor` version `1.0.0` to the toolset of the vendor `arduino`. From now on, we can uniquely refer to this monitor with the pair `PACKAGER` and `MONITOR_NAME`, in this case `arduino` and `ble-monitor` respectively.
```

The compressed archive of the monitor must contain only a single executable file (the monitor itself) inside a single root folder. This is mandatory since the CLI will run this file automatically when a monitor instance is requested.

#### Monitor tools integration

> Each core platform must refer to the specific monitor tools they need by adding them (together with the pluggable discoveries...) inside the `discoveryDependencies` field of the `package_index.json`:

Reviewer: ```suggestion
Each core platform must refer to the specific monitor tools they need by adding them (together with the pluggable discoveries...) inside the `discoveryDependencies` field of the Arduino package index:
```
See my justification above.

> Each core platform must refer to the specific monitor tools they need by adding them (together with the pluggable discoveries...) inside the `discoveryDependencies` field of the Arduino package index:

Reviewer 2: I don't think this is feasible. If both discoveries and monitors are stored in the same list of how do we differentiate them?

            },
            {
              "packager": "arduino",
              "name": "CMSIS",
              "version": "4.5.0"
            },
            ...
          ],
+         "discoveryDependencies": [  <--- Discoveries AND monitors used in the platform
            {
              "packager": "arduino",
              "name": "ble-discovery"
            },
+           {
+             "packager": "arduino",
+             "name": "ble-monitor"
+                                     <--- Version is not required!
+           }
          ]
        },

        {
          "name": "Arduino SAMD Boards",
          "architecture": "samd",
          "version": "1.6.18",
          ...
          "toolsDependencies": [ ... ],
          "discoveryDependencies": [ ... ]
        }

      ],
      "tools": [
        {
          "name": "arm-none-eabi-gcc",
          "version": "4.8.3-2014q1",
          "systems": [ ... ]
        },
        {
          "name": "ble-discovery",
          "version": "1.0.0",
          "systems": [ ... ]
        },
        {
          "name": "ble-monitor",
          "version": "1.0.0",
          "systems": [ ... ]
        }
      ],
    }
  }
}
```

> Adding the needed monitor tools in the `discoveryDependencies` allows the CLI to install them together with the platform. Also, differently from the other `toolsDependencies`, the version is not required since it will always be used the latest version available.

Reviewer: ```suggestion
Adding the needed monitor tools in the `discoveryDependencies` allows the CLI to install them together with the platform. Also, differently from the other `toolsDependencies`, the version is not required since the latest version available will always be used.
```

> Finally, to bind a monitor to a protocol, we must also declare in the `platform.txt` that we want to use that specific monitor tool for that specific protocol with the direcive:

Reviewer: ```suggestion
Finally, to bind a monitor to a protocol, we must also declare in the `platform.txt` that we want to use that specific monitor tool for that specific protocol with the directive:
```

```
monitor.required.PROTOCOL=PLATFORM:MONITOR_NAME
```

the platform can support as many protocols as needed:

```
monitor.required.PROTOCOL1=PLATFORM:MONITOR_NAME1
monitor.required.PROTOCOL2=PLATFORM:MONITOR_NAME2
...
```

in our specific example the directive should be:

```
monitor.required.ble=arduino:ble-monitor
```

where `ble` is the port protocol identification returned by the matching pluggable discovery.

#### Using a monitor tool made by a 3rd party

> A platform developer may opt to depend on a monitor tool developed by a 3rd party instead of writing and maintaining his own.

Reviewer: ```suggestion
A platform developer may opt to depend on a monitor tool developed by a 3rd party instead of writing and maintaining their own.
```

Since writing a good-quality cross-platform monitor tool is very hard and time consuming, we expect this option to be the one used by the majority of the developers.

#### Direct monitor tool integration (not recommended)

> A monitor tool may be directly added to a platform, without passing through the `discoveryDependencies` in `package_index.json`, using the following directive in the `platform.txt`:

Reviewer: ```suggestion
A monitor tool may be directly added to a platform, without passing through the `discoveryDependencies` in the Arduino package index, using the following directive in the `platform.txt`:
```

```
monitor.pattern.PROTOCOL=MONITOR_RECIPE
```

where `MONITOR_RECIPE` must be replaced by the command line to launch the monitor tool for the specific `PROTOCOL`. An example could be:

```
monitor.pattern.custom-ble="{runtime.tools.my-ble-monitor.path}/my-ble-monitor" -H
```

in this case the platform provides a new `custom-ble` protocol monitor tool and the command line tool named `my-ble-monitor` is launched with the `-H` parameter to start the monitor tool. In this case the command line pattern may contain any extra parameter in the formula: this is different from the monitor tools installed through the `discoveryDependencies` field that must run without any command line parameter.

This kind of integration may turn out useful:

> - during the development of a platform (because providing a full `package_index.json` may be cumbersome)

Reviewer: ```suggestion
- during the development of a platform (because providing a full package index may be cumbersome)
```

> - if the monitor tool is specific for a platform and can not be used by 3rd party

Reviewer: ```suggestion
- if the monitor tool is specific to a platform and can not be used by 3rd party
```

Anyway, since this kind of integration does not allow reusing a monitor tool between different platforms, we do not recommend its use.

#### built-in monitor tools and backward compatibliity considerations

Some monitor tools like the Arduino `serial-monitor` or the Arduino `network-monitor` must be always available, so they will be part of the `builtin` package and installed without the need to be part of a real package (`builtin` is a dummy package that we use to install tools that are not part of any platforms like `ctags` for example).

If a platform requires the builtin monitor tools it must declare it with:

```
monitor.required.serial=builtin:serial-monitor
monitor.required.network=builtin:network-monitor
```

> For backward compatibility, if a platform does not declare any discovery or monitor tool (using the `discovery.*` or `monitor.*` properties in `platform.txt` respectively) it will automatically inherits `builtin:serial-monitor` and `builtin:network-monitor` (but not other `builtin` monitor tools that may be possibly added in the future). This will allow all legacy non-pluggable platforms to migrate to pluggable monitor without disruption.

Reviewer: ```suggestion
For backward compatibility, if a platform does not declare any discovery or monitor tool (using the `discovery.*` or `monitor.*` properties in `platform.txt` respectively) it will automatically inherit `builtin:serial-monitor` and `builtin:network-monitor` (but not other `builtin` monitor tools that may be possibly added in the future). This will allow all legacy non-pluggable platforms to migrate to pluggable monitor without disruption.
```
